convert2rhel/unit_tests/utils_test.py

```diff
@@ -874,7 +888,312 @@ def communicate(self):
         pass
 
     def poll(self):
-        return 0
+        return self.return_code
+
+    def close(self):
+        pass
+
+
+@pytest.mark.parametrize(
+    ("output", "skip_stdout_decode", "expected"),
+    (
+        (DummyPopenOutput([u"test".encode()]), True, b"test"),
+        (DummyPopenOutput([u"testê café".encode("utf-8")]), False, u"testê café"),
+        (DummyPopenOutput([u"testê café".encode("utf-8")]), False, u"testê café"),
+        (DummyPopenOutput([u"test".encode()]), False, "test"),
+    ),
+)
+def test_normalize_subprocess_output(output, skip_stdout_decode, expected):
+    result = utils._normalize_subprocess_output(output.stdout.readline, skip_stdout_decode)
+    assert result == expected
+
+
+class TestRunPipedSubprocess:
+    @pytest.mark.parametrize(
+        ("cmd1", "cmd2", "expected"),
+        (
+            (
+                "echo hi",
+                None,
+                "cmd1 should be a list, not a str.",
+            ),
+            (
+                None,
+                "echo hi",
+                "cmd2 should be a list, not a str.",
+            ),
+        ),
+    )
+    def test_run_piped_subprocess_type_error(self, cmd1, cmd2, expected):
+        with pytest.raises(TypeError, match=expected):
+            utils.run_piped_subprocess(cmd1, cmd2)
+
+    @pytest.mark.parametrize(
+        (
+            "cmd1",
+            "cmd2",
+            "process_stdout",
+            "process_stderr",
+            "output_stdout",
+            "output_stderr",
+            "output_return_code",
+            "expected",
+        ),
+        (
+            pytest.param(
+                ["echo", "test"],
+                ["tail"],
+                [u"test".encode()],
+                [u"".encode()],
+                [u"test".encode()],
+                [u"".encode()],
+                0,
+                ("test", "", 0),
+                id="check-correct-success-output",
+            ),
+            pytest.param(
+                ["echo", "test"],
+                ["tail"],
+                [u"".encode()],
+                [u"test".encode()],
+                [u"".encode()],
+                [u"test".encode()],
+                1,
+                ("", "test", 1),
+                id="check-correct-failure-output",
+            ),
+        ),
+    )
+    def test_run_piped_subprocess(
+        self,
+        cmd1,
+        cmd2,
+        process_stdout,
+        process_stderr,
+        output_stdout,
+        output_stderr,
+        output_return_code,
+        expected,
+        monkeypatch,
+    ):
+        process_mock = DummyPopenOutput(process_stdout, process_stderr)
+        output_mock = DummyPopenOutput(output_stdout, output_stderr, output_return_code)
+
+        monkeypatch.setattr(utils.subprocess, "Popen", mock.Mock(side_effect=[process_mock, output_mock]))
+        out, err, return_code = utils.run_piped_subprocess(cmd1, cmd2)
+
+        assert (out, err, return_code) == expected
```

Comment:
I think I would use less mocking in this test.  Yes, it's not a true unittest if we get rid of the mocking, but I think it's a better test overall.

The heart of what makes `run_piped_subprocess()` different from `run_subprocess()` is that it runs two processes, piping the output from the first to the second.  Injecting mocked output as the output from the second command means that we are not actually testing that the communication between the two processes is working.

```suggestion
    @pytest.mark.parametrize(
        (
            "cmd1",
            "cmd2",
            "expected_return_code",
            "expected",
        ),
        (
            pytest.param(
                ["echo", "first line\ntest line"],
                ["tail", "-1"],
                (u"test line", "", 0),
                id="check-correct-success-output",
            ),
            pytest.param(
                ["echo", "test"],
                ["tail"],
                [u"".encode()],
                [u"test".encode()],
                [u"".encode()],
                [u"test".encode()],
                (u"", u"test", 1),
                id="check-correct-failure-output",
            ),
        ),
    )
    def test_run_piped_subprocess(
        self,
        cmd1,
        cmd2,
        expected,
        monkeypatch,
    ):
        out, err, return_code = utils.run_piped_subprocess(cmd1, cmd2)

        assert (out, err, return_code) == expected
```
Note: I didn't test that my suggestion above works.  I just wanted to write it out since I'm not sure if my explanation was clear.

Reply:
Not resolving this comment, but we have this mapped under the gist mentioned in a comment here, so should be fine to resolve it anyway.